ice-adapter/src/main/java/com/faforever/iceadapter/AsyncService.java

```diff
@@ -0,0 +1,85 @@
+package com.faforever.iceadapter;
+
+import com.faforever.iceadapter.exception.AsyncException;
+import lombok.experimental.UtilityClass;
+
+import java.util.concurrent.Callable;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.locks.Lock;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+@UtilityClass
+public class AsyncService {
```

Comment:
I don't understand the purpose of this class. It seems like it would be better to create the ExecutorService and then inject it where needed. Then just use the *Async variants of the completableFuture api directly. I don't think this level of indirection is necessary.

Reply:
It would be great to inject anything anywhere, but that requires up to decrease the STATIC EVERYTHING level a lot :D

Until then the `UtilityClass` is an interesting workaround, that can then later be moved to dependency injection. A similar approach can be seen in #64 .

Reply:
Then maybe instead of injecting I would suggest something more like an ExecutorServiceHolder that doesn't then duplicate the CompletableFutureAPI

Reply:
Whatever I call this class, it will remain static. 
There are 2 options:
1) To put your own ExecutorService in each class, which seems ugly to me.
2) Create a static class that does not carry business logic, performs one task - performs asynchronous tasks in CompletableFuture using virtual threads. Among the advantages, we have control over asynchronous tasks in one place and can easily control them.

Reply:
But we are not doing anything with the tasks besides running them. Then additionally duplicating a standard API. I would think it would be better if this was just a holder that configures the executorService that we then use to populate the ExecutorService of the classes that need it. Then those classes can properly define the execution of their tasks and have full control over any chaining etc.

Reply:
It would also be more portable to dependency injection since you would just replace the assignment of the executor from the holder.

Reply:
Can you give an example as you suggest. So that it works with dependency injection later.

Reply:
Like in the IceAdapter you would have
```
private final Executor executor = ExecutorHolder.getExecutor();
```

Then later
```
clientFuture.thenAcceptAsync(() -> doStuff(), executor)
```

to ensure that the tasks runs with the correct executor. Mostly this is coming from my experience with the client where passing around callbacks just to be run by standard APIs were a little cumbersome. Additionally since the AsyncService still returns completable futures it would be tempting to just use the api of the returned object which goes against the usage of the static AsyncServiceClass.

Reply:
updated